frontend/expr_inferrer.py

```diff
@@ -0,0 +1,372 @@
+            - infer_unary_operation
+            - infer_subscript
+            - infer_compare
+"""
+
+import ast, sys
+from i_types import *
+from context import Context
+from abc import ABCMeta, abstractmethod
+from exceptions import HomogeneousTypesConflict
+
+def infer_numeric(node):
+    """Infer the type of a numeric node"""
+    if type(node.n) == int:
+        return TInt()
+    if type(node.n) == float:
+        return TFloat()
+
+def infer_list(node, context):
+    """Infer the type of a homogeneous list
+
+    Returns: TList(Type t), where t is the type of the list elements
+    """
+    if len(node.elts) == 0:
+        return TList(TNone())
+    list_type = infer(node.elts[0], context)
+    for i in range(1, len(node.elts)):
+        cur_type = infer(node.elts[i], context)
+        if list_type.is_subtype(cur_type): # get the most generic type
+            list_type = cur_type
+        elif not cur_type.is_subtype(list_type): # make sure the list is homogeneous
+            raise HomogeneousTypesConflict(list_type.get_name(), cur_type.get_name())
+    return TList(list_type)
+
+def infer_dict(node, context):
+    """Infer the type of a dictionary with homogeneous key set and value set
+
+    Returns: TDictionary(Type k_t, Type v_t), where:
+            k_t is the type of dictionary keys
+            v_t is the type of dictionary values
+    """
+    if len(node.keys) == 0:
+        return TDictionary(TNone(), TNone())
+    keys = node.keys
+    values = node.values
+    keys_type = infer(keys[0], context)
+    values_type = infer(values[0], context)
+
+    for i in range(1, len(keys)):
+        cur_key_type = infer(keys[i], context)
+        cur_value_type = infer(values[i], context)
+
+        if keys_type.is_subtype(cur_key_type): # get the most generic keys type
+            keys_type = cur_key_type
+        elif not cur_key_type.is_subtype(keys_type): # make sure the dict key set is homogeneous
+            raise HomogeneousTypesConflict(keys_type.get_name(), cur_key_type.get_name())
+
+        if values_type.is_subtype(cur_value_type): # get the most generic values type
+            values_type = cur_value_type
+        elif not cur_value_type.is_subtype(values_type): # make sure the dict value set is homogeneous
+            raise HomogeneousTypesConflict(values_type.get_name(), cur_value_type.get_name())
+    return TDictionary(keys_type, values_type)
+
+def infer_tuple(node, context):
+    """Infer the type of a tuple
+
+    Returns: TTuple(Type[] t), where t is a list of the tuple's elements types
+    """
+    tuple_types = []
+    for elem in node.elts:
+        elem_type = infer(elem, context)
+        tuple_types.append(elem_type)
+
+    return TTuple(tuple_types)
+
+def infer_name_constant(node):
+    """Infer the type of name constants like: True, False, None"""
+    if node.value == True or node.value == False:
+        return TBool()
+    elif node.value == None:
+        return TNone()
```

Comment:
IMO it's a good idea in general (not sure if it applies here) to have an else branch at the end that catches any unexpected input and raises an exception; otherwise, you just implicitly return None here and might get weird behavior later because of that.